### DISCUSSION TITLE
Integrations and extensions

### DISCUSSION THREAD

**Discussion by yilu1015** [terrastruct/d2#345] [Ideas]
Echoing #320, I write to ask if there are plans to offer integration with other tools? As an academic, I write a lot of diagrams in presentations and publications and would be most interested in the following integrations:

- Pandoc, via custom filters (similar to [mermaid-filter](https://github.com/raghur/mermaid-filter))
- Presentation frameworks: RevealJS (like [reveal.js-mermaid-plugin](https://github.com/ludwick/reveal.js-mermaid-plugin) or [Marp](https://github.com/marp-team/marp)
- Documentation generators, such as Gitbook, MkDocs, and Docusaurus. 

Thank you for considering!

**Comment by alixander** [comment]
hey @yilu1015 . yup, we have a dedicated engineer working full time on just integrations. we've just wrapped up https://github.com/terrastruct/d2-obsidian and are getting started on a Slack one. 

Which one would be the top of the wishlist for you?

**Reply by woile** [reply]
Is there a documented way to use from other languages? I'd like to give a try at mkdocs using python

**Reply by alixander** [reply]
What's the use case? Options I can think of:

1. Make calls to `d2` binary (needs to be installed on machine).
2. Make calls to an API (needs internet connection).
3. It looks like you can expose Go code to C, and Python can call C code?? (https://medium.com/analytics-vidhya/running-go-code-from-python-a65b3ae34a2d). idk much about this, but D2 is pure Go, so that should work.

**Reply by woile** [reply]
I don't know how it works on go, but with rust is quite straightforward to build the bindings using [pyo3](https://github.com/PyO3/pyo3). The goal would be to run `pip install d2` and all necessary code should be there.

Option 3 I like the most. The only thing I was able to find was a package [gopy](https://github.com/go-python/gopy) and this article https://blog.gopheracademy.com/advent-2015/gopy/

But again, I don't know much about go bindings.

I think if bindings to more languages were provided, it would be easier for people to build the extensions to their tools themselves. For example, with the python package, it could be integrated with [pymdown](https://facelessuser.github.io/pymdown-extensions/), a lot of the ecosystem would benefit: mainly blog generators and mkdocs.

**Reply by simkimsia** [reply]
I'm sorry @woile and @alixander for jumping in. I just signed up for D2studio pro account after i discover D2 lang.

i am mostly a web developer using python. Would also be interested to find out how to use D2 in python.

Is there any community led effort to make D2 available for python devs?

**Reply by simkimsia** [reply]
@alixander 

> Which one would be the top of the wishlist for you?

do you have a place for us to vote up our preferred integrations?

**Reply by simkimsia** [reply]
Just to add I found unofficial python to d2 diagram

https://github.com/MrBlenny/py-d2

**Reply by simkimsia** [reply]
List of plugins https://terrastruct.com/plugins/ where I found the pythons to d2

**Comment by yilu1015** [comment]
I am an Obisidian user and greatly appreciate the new plugin! Additional integrations would help complete my workflow. Since Pandoc is the swiss knife for document conversion and can output to so many formats (PDF, HTML, Powerpoint, etc.), I think it would be a very good place to start. Thanks! 👍

**Comment by inflation** [comment]
It would be great to use`d2` in a `remark` plugin, so we can add diagrams into Markdown without a separate step.

**Comment by soma-m** [comment]
Docusaurus support would indeed be awesome  to have 😊🔥

**Comment by micmejia** [comment]
anyone here using d2 in docusaurus? https://github.com/facebook/docusaurus/discussions/9608
